diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a1637a..41e63d7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,7 +17,8 @@ find_package(Pthread REQUIRED)
 
 if (NOT DEFINED RIME_TARGET)
     pkg_check_modules(Rime REQUIRED IMPORTED_TARGET "rime>=1.0.0")
-    set(RIME_TARGET PkgConfig::Rime)
+    add_definitions(-I "/tmp/fcitx5/include")
+    set(RIME_TARGET $<LINK_LIBRARY:WHOLE_ARCHIVE,"/tmp/fcitx5/lib/librime.a"> "-L/tmp/fcitx5/lib" leveldb yaml-cpp marisa opencc glog lua)
 endif()
 
 if ("${Rime_VERSION}" VERSION_LESS "1.10.0")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0854a9a..915977f 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -23,6 +23,11 @@ else()
 endif()
 
 add_library(rime MODULE ${RIME_SOURCES})
+target_link_options(rime PUBLIC
+  LINKER:-U,_f5m_is_linear_layout
+  LINKER:-U,_f5m_is_vertical_rl
+  LINKER:-U,_f5m_is_vertical_lr
+)
 target_link_libraries(rime ${RIME_LINK_LIBRARIES})
 install(TARGETS rime DESTINATION "${CMAKE_INSTALL_LIBDIR}/fcitx5")
 fcitx5_translate_desktop_file(rime.conf.in rime.conf)
